xarray/core/common.py

```diff
@@ -577,6 +577,25 @@ def where(self, cond, other=None, drop=False):
 
         return outobj._where(outcond)
 
+    def to_dict(self):
+        """Convert xarray object to a dictionary following xarray naming conventions.
+        """
+        d = {'coords': {}, 'attrs': dict(self.attrs), 'dims': list(self.dims)}
+
+        for k in self.coords:
+            d['coords'].update({k: {'data': self[k].values.tolist(),
+                                    'dims': list(self[k].dims),
+                                    'attrs': dict(self[k].attrs)}})
+        if hasattr(self, 'data_vars'):
```

Comment:
These sorts of checks are best avoided, if possible. I would simply write two implementations of `to_dict`, one on `Dataset` and one on `DataArray`.